### ISSUE TITLE
Writer IDs, Sequence Numbers, Deletes, Replication, Consistency and you

### ISSUE THREAD

**Issue by pauldix** [influxdata/influxdb_iox#709]
This issue is an attempt to kick off some discussion on what kinds of guarantees we need to provide for IOx when it comes to things like linearalizability and reading your own writes. This touches on some of the issues brought up in the discussion on WAL recovery in #651.

Generally we're working with the following constraints:

* Each IOx server within any collection of connected servers requires a unique u32 identifier
* IOx servers should not require distributed consensus or coordination (other than the assignment of a unique ID)
* Individual writes and deletes get assigned a writer ID and a sequence number
* Sequence numbers are always increasing
* We currently have sequence numbers per database (should this be changed to just a global?)

The important parts to this are the WAL Buffer, the Mutable Buffer, and the Catalog.

The WAL Buffer organizes data into immutable segments that get persisted once closed. Those segments have a monotonically increasing integer id. Segments are also associated with a writer ID (segments can have writes and deletes from many different writers). A segment will have many individual writes, each with its own writer id and sequence number. The segment id exists only as a way to index the WAL so that readers can keep track of which segment they last read from.

The Mutable Buffer organizes data into partitions and within those into chunks. Partitions have non-overlapping data while chunks can have overlapping data. Chunks within a partition have monotonically increasing integer ids. Once closed and converted over to the Read Buffer and/or Parquet files in object store, a chunk is immutable.

The Catalog is persisted to the object store and keeps track of chunks that have been written along with their summary stats. In addition to schema information, the chunk summary includes what writer IDs have data in the chunk and what their min and max sequence numbers are. It will also include, if applicable, what writer IDs have segments in the chunk and what the min and max segment IDs are.

## Synchronous Lifecycle
When a write or delete comes into an IOx server it gets converted into a Flatbuffers object, which has the writer ID and a sequence number. From there one or more of the following can happen, but each is optional:
* The write/delete goes into the WAL buffer
* The write/delete goes into the mutable buffer
* The write/delete is replicated to other IOx servers synchronously (during the request lifecycle). When the request lands here, the same options are available.

## Asynchronous Lifecycle
Periodically, operations will be performed outside the write/delete request/response lifecycle. These are for persisting data to object storage.
* Write a WAL Segment (collection or writes & deletes) to object storage
* Write a Chunk to object storage, which in turn updates the Catalog

## Do we need serializability?
For insert only workloads (which is most of what we do), ordering writes doesn't really matter. We only have an issue with ordering writes if writers are updating existing records (with different data) or performing deletes and subsequent inserts into a range the delete would have touched.

We don't need to optimize for updates or even ensure serializability with them. We just need to ensure that we're able to bring multiple servers into some consistent state eventually.

## All-in-one scenario

Consider the following scenario where the writer and sequence assignment, the wal buffer and the mutable buffer all happen in the same process (writer 1).

```
┌─────────┐ 
│Writer 1 │ 
└────┬────┘ 
     │      
     │      
┌────▼─────┐
│WAL Buffer│
└───┬──────┘
    │       
    │       
 ┌──▼─────┐ 
 │Mutable │ 
 │Buffer  │ 
 └────────┘ 
```

Writes should land in the WAL Buffer and the Mutable Buffer in order. On recovery of the process, the WAL Segments from object storage should be played back into the Mutable Buffer in order. Note that a WAL Buffer and Mutable Buffer exists for each database.

## An example setup with multiple servers

IOx can be configured to split these roles out across multiple servers. A smaller version of our own production environment will likely have this look:

```
     ┌─────────┐       ┌─────────┐      
     │Writer 1 │       │Writer 2 │      
     └─────────┘       └─────────┘      
          │ │             │ │           
          ├─┴─────────────┴─┤           
          │                 │           
          ▼                 ▼           
    ┌───────────┐     ┌───────────┐     
    │WAL Buffer │     │WAL Buffer │     
    │Writer 3   │     │Writer 4   │     
    └───────────┘     └───────────┘     
          │                 │           
          │                 │           
        ┌─┘                 └──┐        
        │                      │        
        │                      │        
        ▼                      ▼        
┌───────────────┐      ┌───────────────┐
│Mutable Buffer │      │Mutable Buffer │
│Writer 5       │      │Writer 6       │
└───────────────┘      └───────────────┘
```

Writers 1, 3 and 5, will exist in the same AZ while writers 2, 4, and 6 exist in another. We'd likely have 3-4 AZs represented.

At the top we have a routing tier that parses incoming writes and deletes and assigns a writer id and sequence number. While still in the request/response lifecycle it will attempt to replicate that data to Writer 3 and 4, which are responsible for buffering writes in memory (the WAL Buffer). The writers at the top send data to other AZs. At that point a success is sent back to the client.

The WAL Buffer will asynchronously send data to the mutable buffer servers. Note that each WAL Buffer sends data only to a single mutable buffer server.

Clients making requests can ask for single, quorum, or full before a success is returned. Because the WAL Buffer servers keep data in memory until it is flushed out to object storage, the replication will determine how many failures can occur before a write is lost.

If a failure occurs at the routing layer, the load balancer in front of it (or a client that knows about both servers) should redirect requests to the other server. The other server will continue to replicate to both WAL Buffer servers.

Because individual requests can fail and because WAL Buffer servers can crash and lose writes, this means that the Mutable Buffer on writers 5 and 6 can have different data. Queries sent to the IOx infrastructure should have sticky sessions so that queries from the same user get sent to the same Mutable Buffer server. Although we still need a way to resolve inconsistencies.

If the Mutable Buffer server restarts, it should be able to rebuild its mutable buffer from the WAL Buffer as discussed in #651. Its API should expose a way to have it make requests to a WAL Buffer server to pick up writes (pull based replication).

For a WAL Buffer server that fails, it can be instructed via API call to request missing data from some other WAL Buffer server (again pull based replication).

Server restarts, additional servers and others that need to pick up data should be managed externally. They should have an API so external agents can tell them to pick up data from any number of other IOx servers or from the object store directly. Once they report completion of the task the agent should be able to instruct the IOx server to start responding to requests.

A consistent view can be built by combining the streams from multiple WAL Buffers, using some rules.

## Multiple writers, buffers, readers

Within our own environment, we'll be operating each of the boxes in the previous diagram on a separate server. We'll also operate multiple servers at each stage. But different setups could have completely different configurations. To highlight the various points at which there might be inconsistency, we'll use the following diagram.

```
     ┌─────────┐       ┌─────────┐      
     │Writer 1 │       │Writer 2 │      
     └─────────┘       └─────────┘      
          │ │             │ │           
          ├─┴─────────────┴─┤           
          │                 │           
          ▼                 ▼           
    ┌───────────┐     ┌───────────┐     
    │WAL Buffer │     │WAL Buffer │     
    │Writer 3   │     │Writer 4   │     
    └───────────┘     └───────────┘     
          │   │         │   │           
          │   │         │   │           
        ┌─┴───┴─────────┴───┴──┐        
        │                      │        
        │                      │        
        ▼                      ▼        
┌───────────────┐      ┌───────────────┐
│Mutable Buffer │      │Mutable Buffer │
│Writer 5       │      │Writer 6       │
└───────────────┘      └───────────────┘
```

At the top with writers 1 and 2 is a simple routing/synchronous replication layer. Write and delete requests come in and are parsed into their Flatbuffers structure and immediately assigned writer id and sequence number. Those FB structs are sent to writers 3 and 4 (via HTTP request) and on success return a response to the client.

From the WAL Buffer servers (writers 3 & 4), requests are made to send their data to the mutable buffer servers (5 & 6). The Mutable Buffer servers will be responsible for deduplicating writes from the buffers. In a real-world scenario we'd likely have a mutable buffer hooked up to only one of the WAL Buffers and have it connect to the other one only on failure of its primary.

### WAL Buffer Ordering

Let's work through the WAL Buffer part first. Here's a sample write sequence where the parts to the left of the arrow represent the writer id and sequence number (e.g. `1,2` writer 1 sequence 2). To the right of the arrow are the writers that receive the write. So the arrow represents a network hop.

```
1,1 -> 3,4 (writer 1 sequence 1 replicate to writer 3 and 4)
1,2 -> 3 (writer 1 sequence 2 replicates to 3, but fails 4)
1,3 -> 4 (writer 1 sequence 3 replicates to 4, but failed 3)
2,1 -> 3 (writer 2, sequence 1 replicates to 3 and starts the next but first)
1,4 -> 3,4 (writer 1, sequence 4 replicates to 3 and 4)
2,1 -> 4 (and writer 2, sequence 1 finishes and now replicates to 4)
```

Given the above sequence we'd end up with two different WAL Buffers for writers 3 and 4 respectively:

```
3:   4:
1,1  1,1
1,2  1,3
2,1  1,4
1,4  2,1
```

Those represent the WAL Buffer from those servers. The individual buffer can be broken out into segments in any number of ways. For example, it could look like this:

```
3:
Segment 1: 1,1
Segment 2: 1,2 - 2,1 - 1,4

4:
Segment 1: 1,1 - 1,3
Segment 2: 1,4
Segment 3: 2,1
```

Again, segment numbers aren't used to merge streams. Although the timestamps of their files may be used to roughly order segments together, the writes within segments will be ordered based on shared writer ids.

A reader pulling only from either 3 or 4 won't have a problem as there's only one ordering represented. A reader pulling from both will obviously get different orderings. The other important thing to note is that the segment numbers on 3 & 4 may not be the same and segments may not have exactly the same writes (even if their ordering is the same). Segments are groups of writes that get closed out by time or size.

A writer ID and sequence number appearing in both WAL buffer servers is the only way to situate the stream of 3 with respect to the stream of 4. If 3 & 4 don't get writes from any of the same writers, their streams can't be ordered, other than by the creation time of the segment.

To resolve the conflict in this stream, we choose the writer with the higher ID as having the correct ordering and we'd need to merge in any writes that one doesn't have. With that algorithm we'd end up with the following stream:

```
1,1
1,2
1,3
1,4
2,1
```

Whatever the operations were for each of these writes (either insert, update to existing value or delete), they should be applied in that order. When reading the WAL from two servers, the segments should be advanced and merged together so that writes go in order from individual writers. This will only be possible if writes from the same writer occur in both servers.

If writes only occur from different writers like:

```
3:   4:
1,1  2,1
1,4  2,2
1,6  2,3
```

We would apply writes in order of segment created date (the file time stamp in object storage) and then writer ID. For example, say the two orderings above existed in one segment for each. Further say that the segment on 4 had an older time stamp. All of 4's writes in that segment would be played before 3. Which means if 1,1 happened to be a delete to all the data, everything written from 4's segment would be cleared out on the replay (even if those writes originally arrived at writers 3 and 4 in a different order in real time).

In this case things like clock drift across servers (or within your object store) may impact ordering.

### Chunk Write Ordering

When data is sent from the WAL Buffer onward to the Mutable Buffer (either in the same server or across servers), the write to the Mutable Buffer will include the writer id from the WAL Buffer and the segment number the write comes from.

When the chunk is saved to object storage, its summary statistics should include:
* Each writer ID with individual writes and the min and max sequence
* If writes came from the WAL Buffer from 1 or more servers, their writer id and the min and max segment number

This information can be used to determine where to replay from WAL segments (and which writes from older segments can be safely ignored).

In the above diagram, it would be possible for writer 5 and 6 to create chunks that are different from each other. This could be because of when a chunk was triggered to close or it could be because the subscription requests from 3 and 4 arrived in different order.

Inconsistency between the two will have to be handled on a partition (ideally after it is no longer receiving writes). We could do this either by replaying the WAL and constructing an ordering based on the rules in the previous section, or if the WAL isn't available, it could simply merge the data from the two servers together. Then if previously deleted data showed up again, the user could issue a delete to clear it out.

## Mutable Buffer, Partitions, Chunks and consistency

Because data is split into partitions, we don't have to check consistency by replaying the WAL from some point. We can also check on individual partitions from multiple writers. Say you have Mutable Buffer servers 5 and 6 from above and the writes coming in were partitioned by hour. Say we have hour `2021-01-29T18:00` and some number of chunks beneath that which have been persisted.

When the chunks are persisted, they should have a signature of the data in them and we should have a signature for the partition. If writers 5 & 6 have the same signature for that partition, we know it's consistent.

If it isn't, we should have a method of combining the data for a partition from two different writers. We union and remove duplicates. This means that a chunk that had processed a delete in a different order than one on the other writer would have data show up that may have previously been deleted. As long as the user has the ability to rewrite this new partition data (a new chunk replacing all old ones) and then send a delete to make a correction, they have everything they need.

To ensure that the mutable buffer on 5 and 6, or at least the partitions they access are consistent, an external process will either have to manage this by pulling the data from the WAL and writing it in a new spot, or checking on the mutable buffers in 5 and 6 periodically. The signatures can be used to determine if they are inconsistent and the operator can take action either by ingesting WAL Buffer copies, or merging the two chunks together.

APIs should exist in IOx to perform any of these repairs on request. But the logic for when to do what and why should be contained in the operating software.

## Ensuring Sequence Numbers are Increasing Across Restarts

Since we don't want to persist the sequence number on every individual write, but we still want to ensure that they're always increasing so that writes from a writer can be ordered, we persist them in blocks. For example, we'll update the catalog every 500,000 sequence numbers.

When the server starts it should look for what the last block was an add the step size * 2. Say we updated when we hit 3,500,000 but the system restarted or crashed at some point after that. On startup we'd see the last value and start from `3,500,000 + (2 * 500,000)` so 4.5M. Starting from twice the step size gives the server a window of time to be able to persist the update to the catalog while updates are still coming in. If the object store is unreachable for 999,999 writes, then the server would have to start rejecting writes until it can update the catalog to ensure that a restart won't overlap.

## Ensuring Segment Numbers are Increasing

Chunks keep track of the min and max segment numbers for each writer that has a WAL Segment in the chunk. This is optional as a chunk can be filled up without ever reading from a WAL, in which case it only has the min and max from the actual writers.

Because a segment can have writes sent to the mutable buffer before it is closed (and thus persisted), we need to ensure that the segment ID doesn't get used if the WAL Buffer restarts before it persists it. We'll use the same trick as before with the sequence numbers, but likely on a smaller scale. Update the catalog every 100 segments and increase the last number by 200 on restart.

## Reading your own writes

To give users the option to read their writes, write requests should include the writer ID and sequence number in the response. The user can optionally pass this information in a read request to ensure that any server answering the query has at least seen a sequence equal to or higher for that writer id.

## Questions

Some questions for discussion:

1. Should we update the mutable buffer with the set to track writer/sequence?
1. Should we have a new component which can take the WAL Segment streams or write streams from multiple writers and create an ordering which can then be fed into the mutable buffer?

## Closing thoughts

Almost all of the time, I wouldn't expect users to be updating previous records with new data. Double writes of the same data certainly may happen, but this can be resolved when a chunk is persisted just by sorting and removing duplicates. I also wouldn't expect deletes against ranges of data that are actively being written to. And even if this is the case, this should be fairly predictable under normal operation.

The main thing I'd like to achieve is something that performs very well under normal operation, but has the flexibility to account for server restarts because of deploys of new versions, moving databases over to new servers, adding read replicas etc. I want to give users the tools to fix any consistency issues that may arise from unforseen circumstances and ideally give them the ability to recover from logs where they want to.

Much of this assumes that we're not dealing with contentious writes and deletes. That is, we're optimized for inserts.

The goal here is not to ensure an exact ordering as the requests came in, but to have some state that can be recovered reliably. IOx is designed for inserts and deletes against ranges where you're not simultaneously writing data. If you need exact ordering guarantees you'll have to construct an infrastructure design that has all writes to a single database come into one writer and WAL Buffer.

**Comment by alamb** [comment]
TLDR I have two primary reactions / points for further discussion:
1. I am not sure this design will lead to eventual consistency
2. I am trying to understand the rationale for writing multiple copies of the data to object store

## Primary Requirement: Eventually consistent reads

This is the key design constraint in my opinion:

> We just need to ensure that we're able to bring multiple servers into some consistent state eventually.

It means (to me) if you stop writes and wait a while, eventually all reads should see the same data. It also means if you then nuke a node and restart it will be able to recover to a point where the reads will see the same data.

Imagine a customer wrote some data to IOx, and went to bed. Now the next day they issue some queries from the web UI  which hits data from one Mutable Buffer and I run a curl command with the same query from my application server that is served from another Mutable Buffer. If the data is not the same, InfludData is going to get a support case.

I can't quite convince myself that the design in this write up will satisfy the eventual consistency requirement, though I may just need to study the details more.

My hangup is that since there is *no database wide ordering of writes* (as both WAL Buffer 3 and WAL Buffer 4 might see writes, updates, and deletes in a different order). I can't figure out how this will lead to an eventually consistent view of the data.

Breaking out the CS theory, in order to get consistency without coordination, [the CALM paper](https://arxiv.org/abs/1901.01930) basically says operations need to be "monotonic." As I understand it, inserts of new data satisfy this property, but updates of existing data and deletes do not. Thus I conclude, in a hand wavy way, that we need some sort of distributed coordination to order the operations in IOx to get eventual read consistency.


## Multiple Peer WAL Buffer Writers vs Primary / Standby

The CALM paper counsels keeping distributed coordination to a minimum. While this is a Captain Obvious type statement, perhaps it could be applied it to IOx with a minor tweak to your pictures with a Primary / Standby system:


```
┌──────────┐           ┌──────────┐
│ Writer 1 │           │ Writer 2 │
└──────────┘           └──────────┘
      │                      │
      └───────────────────┐ ┌┘
                          │ │
                          ▼ ▼
┌──────────┐          ┌──────────┐
│WAL Buffer│          │WAL Buffer│
│ Writer 3 │◀─────────│ Writer 4 │
└──────────┘          └──────────┘
      │                     │
      │                     │
      │                     │
      ▼                     ▼
┌──────────┐          ┌──────────┐
│ Mutable  │          │ Mutable  │
│  Buffer  │          │  Buffer  │
│ Writer 5 │          │ Writer 6 │
└──────────┘          └──────────┘
```

In this case all writes for a single database would be sent to a "primary" WAL buffer which would determine order as well as decide when the write had been sufficiently replicated to respond OK to the client. The other WAL buffers would be responsible for buffering the segments and writing them to object storage as a safeguard against a primary failure.

Obviously in this case, distributed consensus would be required, but ONLY to determine which server was the primary, and there would be a period of time during failover where write latency was higher; However, primary / secondary failover is a well studied problem and there are ample implementations to draw inspiration from.


## Usecase for Multiple Writers (of parquet chunks) to object store?
Another thing I was wondering was what is the usecase for multiple writers of the same parquet chunk data to object storage? This property is a significant source of complexity in the design I think.

The only usecase I can think of is running in multiple AZs - so a copy of the data is available if one AZ fails. Keeping a replica of the data in another AZ could also reduce cross-AZ network traffic while querying data.

However, given object storage itself is already highly redundant and durable, writing the same data there twice seems like a very expensive way for multi-AZ support. I am not really thinking in terms of storage byte cost, but in terms of CPU / memory needed to re-encode the same data as parquet, compact it, etc).

Given object storage is immutable in our design, maybe we could just have one MutableBuffer writer to object store and copy the files that were saved. Perhaps this is what you had in mind already

BTW the same reasoning applies to writing redundant WAL buffer segments to object storage too -- once one WAL buffer has written its data to object storage, it seems like there very little reason for a second WAL buffer to go through the same trouble, though perhaps handling the AZ failure is important in this case.

**Comment by pauldix** [comment]
Any of the configurations you talk about is possible with the design. A single WAL Buffer server, a single server writing Parquet files (which was what I was thinking there). Two servers holding the Mutable Buffer was to ensure there was another copy immediately queryable, but isn't required. But I wanted to call out the two spots where we might need to do consistency checks later, first in the WAL and later with the partitions.

I think there's a tradeoff here between consistency and availability. I was thinking that in the near term, availability is the most important thing, while consistency can be guaranteed on a longer time frame.

Take the two WAL Buffers example. Say we're partitioning data into 1 hour blocks. After each 1 hour block passes, you can have an IOx server (not in that picture) process the WAL streams from the two servers, producing a unified ordering and writing out the partition to object storage. Then you tell the Mutable Buffer servers to reload their buffers from the unified object storage (only if the signatures differ, no point in doing the work if we don't have to). So we ensure availability and (in most cases) they're consistent, but we catch the edges after a partition closes. Fix it in post, as they say.

The goal with having the two servers write their segments to object storage is to make sure we don't lose writes. We could configure one of the WAL servers to not persist its segments and keep it up only as an in-memory standby. My thought was that it would be better to have the servers persisting data without external coordination, that way if something goes off with Conductor, they would at the very least continue to write data to object storage so it could be replayed and recovered later.

The thing I want to figure out in this issue is this: if we have this functionality, can we create an architecture that has the properties we want? Part of this is that I don't think there's a one size fits all, so I think we should be able to modify it on a per database basis.

**Comment by alamb** [comment]
@pauldix  -- what you are saying makes sense. I see it now -- basically as long as (only) one of the mutable buffer servers produces the stream which is eventually turned into parquet chunk files there is the single authoritative ordering I was looking for.

I think I was confused as I was assuming  the mutable buffer servers were also (eventually) both producing parquet chunk files

In terms of two WAL buffer servers persisting the stream of writes, that seems reasonable to me too for the extra durability benefits you suggest, though I am still struggling to understand what long term value the stream of writes written into the WAL buffer but not persisted to parquet is. I understand the short term value of keeping a read replica up to date prior to the chunks being persisted

**Comment by alamb** [comment]
Things I think would help this design / description of replication:
* Giving a name to the stream of writes a particular WAL buffer sees (and the segment files that store that stream)
* Describing recovery always in terms of a particular stream (so the recovery story when the stream is already persisted to object store, as well as what to do when the stream was only partially persisted but there is another distinct stream available)

It might be worth also considering ways to avoid reconciling / merging distinct streams of writes at all initially if possible and focus on slower but simpler recovery mechanisms -- I have thoughts here but they are not yet fully formed

**Comment by e-dard** [comment]
I've given this a read, and I think I need to give it a few more reads :-) 

I worry about using things like sticky sessions to avoid perceived data-flapping. Session are tied to clients and I don't think it's unreasonable that a user would have multiple clients on the go. @alamb mentions this too

**Comment by domodwyer** [comment]
After working through the above and discussing internally, we've produced a working document detailing the replication & consistency design we expect to implement: 

[IOx Replication & Eventual Sequential Consistency](https://docs.google.com/document/d/1q31f-GCJWYv8s9m6dSVKivCZ6c0XU9Ozi45uC0T1O0o/edit?usp=sharing)

The design is largely there, but plans change! I'll keep the document up to date as things progress.

**Comment by jacobmarble** [comment]
@pauldix can this be closed?

**Comment by pauldix** [comment]
Yeah, no longer applicable
